### ISSUE TITLE
Create full identity UI source - AddDefaultTokenProviders missing

### ISSUE THREAD

**Issue by pholly** [aspnet/Docs#8435]
Unless I'm missing something, `AddDefaultTokenProviders` should be added to the Identity configuration in the [Create full identity UI source](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/scaffold-identity?view=aspnetcore-2.1&tabs=visual-studio#create-full-identity-ui-source) section. Otherwise, the app will throw an error when generating and storing tokens, like during Forgot password / recover password. `AddDefaultTokenProviders` configures a default DataProtectionTokenProvider.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 804e24dd-7e30-8cf5-2c99-1ffacb4582f6
* Version Independent ID: 3e25547b-e2c9-6634-2c38-2a77207f17d4
* Content: [Scaffold Identity in ASP.NET Core projects](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/scaffold-identity?view=aspnetcore-2.1&tabs=visual-studio#create-full-identity-ui-source)
* Content Source: [aspnetcore/security/authentication/scaffold-identity.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/security/authentication/scaffold-identity.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
@HaoK  please confirm.

**Comment by Rick-Anderson** [comment]
@pholly  I just tested this and it works fine without calling `AddDefaultTokenProviders`

**Comment by pholly** [comment]
@Rick-Anderson I just uncommented it from my code and ForgotPassword give this error. There is no other place in the app where token providers are registered.
![image](https://user-images.githubusercontent.com/3717650/45182144-7e0e4f80-b1ee-11e8-850f-bcc5d8c3c083.png)

**Comment by Rick-Anderson** [comment]
@pholly I just tested and set a break-point in `public async Task<IActionResult> OnPostAsync()` in *Areas\Identity\Pages\Account\ForgotPassword.cshtml.cs* and ` var code = await _userManager.GeneratePasswordResetTokenAsync(user);` works fine.

**Comment by pholly** [comment]
@Rick-Anderson no token providers are registered so I don't see how you're not getting an error. Unless you're working on a more recent code base that registers token providers without calling AddDefaultTokenProviders.

This is in `UserManager.cs`:

```c#
foreach (var providerName in Options.Tokens.ProviderMap.Keys)
                {
                    var description = Options.Tokens.ProviderMap[providerName];
                    
                    var provider = (description.ProviderInstance ?? services.GetRequiredService(description.ProviderType)) 
                        as IUserTwoFactorTokenProvider<TUser>;
                    if (provider != null)
                    {
                        RegisterTokenProvider(providerName, provider);
                    }
                }
```

There must be some place in your code that adds token providers to the ProviderMap. `AddIdentity` doesn't.

**Comment by HaoK** [comment]
AddDefaultIdentity is AddIdentity + AddDefaultTokens + AddDefaultUI

**Comment by pholly** [comment]
Yes this issue is with `AddIdentity`. The example in the docs I referenced above uses `AddIdentity` without `AddDefaultTokenProviders` so no token providers are registered. 

I searched the Identity repository and the only place that calls `AddTokenProvider` is `AddDefaultTokenProviders` so I'm not sure why @Rick-Anderson is not getting the error.

**Comment by Rick-Anderson** [comment]
I see the problem. When you scaffold full Identity it uses *Areas\Identity\IdentityHostingStartup.cs*

```
public class IdentityHostingStartup : IHostingStartup
{
    public void Configure(IWebHostBuilder builder)
    {
        builder.ConfigureServices((context, services) => {
            services.AddDbContext<RPauthAllScaffoldContext>(options =>
                options.UseSqlServer(
                    context.Configuration.GetConnectionString("RPauthAllScaffoldContextConnection")));

            services.AddDefaultIdentity<MyUser>()
                .AddEntityFrameworkStores<RPauthAllScaffoldContext>();
        });
    }
}
```

I'll get the doc updated.

**Comment by pholly** [comment]
Yeah you're supposed to delete that file and add `AddIdentity` and other options to your Startup.cs

**Comment by Rick-Anderson** [comment]
>Yeah you're supposed to delete that file and add `AddIdentity` and other options to your *Startup.cs*

@pholly  where does it say that. I seem to remember writing that but I can't see it in the doc.

**Comment by Rick-Anderson** [comment]
@pholly thanks for persisting. I'll get this updated this week.

**Comment by pholly** [comment]
@Rick-Anderson Funny after I wrote that I searched the Scaffolding doc but didn't find it then remembered it's in the migration doc here: https://docs.microsoft.com/en-us/aspnet/core/migration/20_21?view=aspnetcore-2.1#update-after-scaffolding-identity

Thanks for writing the docs and responding to feedback!

Regarding deleting `Areas/Identity/IdentityHostingStartup.cs`, perhaps Scaffolding.txt and comments in `IdentityHostingStartup.cs` could describe what that file is for and when it should be deleted. It seems if someone already has Identity configuration code in `Startup.cs` that file should be deleted.

Thanks!

**Comment by KenSpr** [comment]
I have read this thread but I'm not clear of the final answer. We are using a custom user and role class but everything else should be stock Identity (nothing scaffolded out). In ConfigureServices, should we include AddDefaultTokenProviders() or leave it out as shown here:
		services.AddIdentity<MyUser, MyUserRole>()
			    .AddEntityFrameworkStores<AppDbContext>()
		   	 //.AddDefaultTokenProviders()
			   .AddDefaultUI();

**Comment by Rick-Anderson** [comment]
Add `AddDefaultTokenProviders`
This is fixed in #8458
